# NAM-genomes :corn: :corn: :corn:

<details>
  <summary>For co-authors (click to expand)</summary>

### NAM genome paper related scripts and workflows

Please follow this structure for your contribution.

1. Add all the scripts for one specific task inside aptly named folder (eg: `agp-generation`), use hyphen for gaps and use all lower-case letters (if possible).

2. Add a README.md file within the folder, specifically addressing the order and inputs required for the scripts you put in the folder

3. If you have any figures, please create a folder within that section folder  (please name it "assets") and drop them there. You can also add other files or things that doesn't qualify as scripts in there.

5. Please check if your section is linked correctly from the main README page

5. Please let me know if you have any questions or have any suggestions!

Go, NAMsters!

![NAMsters](/assets/NAMster2.png)

</details>


# Table of contents:

1.  Genome assembly methods ([contig, hybrid scaffold](genome-assembly/README.md) and [pseudomolecule construction](agp-generation/README.md) )
2. [Quality control (raw data: Illumina, PacBio, RNA-Seq; and Assembly)](quality-control/README.md)
3. Gene prediction and annotation ([evidence-based](gene-prediction/evidence-based.md), [BRAKER](gene-prediction/ab-initios.md), [merging annotations](gene-prediction/merge-predictions.md))
4. [Centromere annotation](centromere_annotation/README.md)
5. [Knob analyses](knob-analyses/README.md)
6. [DNA methylation and Accessible Chromatin Regions (ACR) analyses](methylation/README.md)
7. [PanGenome analyses](pangenome-analyses/README.md)
8. [Transposable Element Annotation](te-annotation/readme.md)
9. [Tandem Repeat Arrays](tandemrepeat-quantification/README.md)
10. [Structural variation detection](structural-variation/README.md)
11. [Analysis of Flowering-time Genes](structural-variation/FTGene_Expression_README.md)
12. [Analyses of disease resistance Genes](NLR-genes/README.md)
13. Population Genetic Analysis ([GERP](gerp/README.md), [abc](abc/README.md))
14. [Genome-wide Association Study and Variance Component Analysis](gwas-vcap/README.md)
15. [Fractionation Analysis](fractionation/README.md)

# Figures/Plots

[Figures and plots used in the publication](publication-plots/README.md)

# Links for external data

1. [CyVerse Data Commons](https://datacommons.cyverse.org/browse/iplant/home/shared/NAM/NAM_genome_and_annotation_Jan2021_release)
2. MaizeGDB [NAM-page](https://www.maizegdb.org/NAM_project), [GenomeBrowser](https://maizegdb.org/gene_center/gene), [FTP site](https://download.maizegdb.org/).
3. [Gramene](http://ftp.gramene.org/nam/release-3/)
3. Raw Datasets (PRJEB31061 and PRJEB32225)
    - [PacBio](https://www.ebi.ac.uk/ena/browser/view/PRJEB31061)
    - [Illumina](https://www.ebi.ac.uk/ena/browser/view/PRJEB31061)
    - [RNAseq: E-MTAB-8633 and E-MTAB-8628](https://www.ebi.ac.uk/ena/browser/view/PRJEB32225)
    - [BioNano](https://www.ebi.ac.uk/ena/browser/view/PRJEB31061)
    - [Methylation data: E-MTAB-10028](https://www.ebi.ac.uk/ena/browser/view/PRJEB31061)
    - [NLR trees](https://itol.embl.de/shared/xCJbI9ndshEK)
 4. [NAM-webpage](https://nam-genomes.org/)

# Citation

Matthew B Hufford, Arun S Seetharam, Margaret R Woodhouse, Kapeel M. Chougule, Shujun Ou, Jianing Liu, William A Ricci, Tingting Guo, Andrew Olson, Yinjie Qiu, Rafael Della Coletta, Silas Tittes, Asher I Hudson, Alexandre P. Marand, Sharon Wei, Zhenyuan Lu, Bo Wang, Marcela K Tello-Ruiz, Rebecca D Piri, Na Wang, Dong won Kim, Yibing Zeng, Christine H Oâ€™Connor, Xianran Li, Amanda M Gilbert, Erin Baggs, Ksenia V Krasileva, John L Portwood II, Ethalinda KS Cannon, Carson M Andorf, Nancy Manchanda, Samantha J Snodgrass, David E Hufnagel, Qiuhan Jiang, Sarah Pedersen, Michael L Syring, David A Kudrna, Victor Llaca, Kevin Fengler, Robert J. Schmitz, Jeffrey Ross-Ibarra, Jianming Yu, Jonathan I Gent, Candice N Hirsch, Doreen Ware, R Kelly Dawe. De novo assembly, annotation, and comparative analysis of 26 diverse maize genomes. bioRxiv, [10.1101/2021.01.14.426684](https://www.biorxiv.org/content/10.1101/2021.01.14.426684v1), 2021. 

# Contacts

1. [R. Kelly Dawe](http://www.dawelab.org/)
2. [Doreen Ware](https://www.cshl.edu/research/faculty-staff/doreen-ware/)
3. [Matthew Hufford](https://mhufford.public.iastate.edu/HuffordLab/home.html)
4. [Candice Hirsch](http://hirschlab.cfans.umn.edu/)


Please check our [contributions](https://github.com/HuffordLab/NAM-genomes/graphs/contributors) page for all the authors who worked on this GitHub documentation (and the sections they worked on!).
